src/datasets.cpp

```diff
@@ -498,6 +499,38 @@ DataFrame DataFrame::copy() const
     return *new_frame;
 }
 
+DataFrame DataFrame::sample(int nrow, int seed) const{
+    // Set random seed for reproducibility if specified
+    if (seed == -1){
+        // obtain a random number from hardware
+        std::random_device rd;
+        seed = rd();
+    }else{
+        assert(seed >= 0);
+    }
+    // number of rows to pull
+    if (nrow == -1){
+        nrow = this->length();
+    }else{
+        assert(nrow > 0);
+    }
+    // Seed the generator
+    std::mt19937 eng(seed);
+    // Draw row indices from uniform distribution
+    std::uniform_int_distribution<> distr(0, this->length()-1);
+    // Create new empty DataFrame
+    DataFrame *new_frame = new DataFrame();
+    // pull random rows (as pointers, not copies) until full
+    while (new_frame->length() < nrow)
+    {
+        // get random row index with replacement
+        int rand_row = distr(eng);
+        // get pointer to that row in original dataframe and store in bootstrap
+        new_frame->addRow(this->row(rand_row));
+    }
+    return *new_frame;
```

Comment:
I think we're needlessly putting the result on the heap (which is what the `new` keyword does).
The `new` approach is necessary when we're returning pointers (the function's local memory doesn't persist when the function ends, so we need the object we're pointing to be on the stack in order for the returned pointer to actually point to something real). But in this case, since the function returns an actual DataFrame instead of a pointer, we don't need to allocate space on the heap -- it'll keep existing in the local memory of whichever scope called the `sample` functions -- assuming I'm understanding heap/stack correctly!)

We could do something like this:
```suggestion
    // Create new empty DataFrame
    DataFrame new_frame = DataFrame();
    // pull random rows (as pointers, not copies) until full
    while (new_frame.length() < nrow)
    {
        // get random row index with replacement
        int rand_row = distr(eng);
        // get pointer to that row in original dataframe and store in bootstrap
        new_frame.addRow(this->row(rand_row));
    }
    return new_frame;
```
but I see that `transpose` and maybe some other functions can be updated, so I can address that in a separate PR.